sympy/assumptions/refine.py

```diff
@@ -240,6 +240,73 @@ def refine_Relational(expr, assumptions):
     return ask(Q.is_true(expr), assumptions)
 
 
+def refine_re(expr, assumptions):
+    """
+    Handler for real part.
+
+    >>> from sympy.assumptions.refine import refine_re
+    >>> from sympy import Q, re
+    >>> from sympy.abc import x
+    >>> refine_re(re(x), Q.real(x))
+    x
+    >>> refine_re(re(x), Q.imaginary(x))
+    0
+    """
+    from sympy.functions.elementary.complexes import re, im
+    arg = expr.args[0]
+    if arg.is_Atom:
+        if ask(Q.real(arg), assumptions):
+            return arg
+        if ask(Q.imaginary(arg), assumptions):
+            return 0
+    if isinstance(arg, Mul):
+        reals = [refine(re(a), assumptions) for a in arg.args]
+        ims = [refine(im(a), assumptions) for a in arg.args]
+        for i in range(len(reals)-1):
+            first_re = reals.pop()
```

Comment:
Why not just traversing the list in opposite direction from may be, from `-2` to `-len(reals)`. Is popping necessary. I can not figure out the reason for doing so.

Reply:
I need to remove the elements that have been already multiplied together, and replace the answer with their product.